sbi/inference/snpe/snpe_base.py

```diff
@@ -350,6 +349,9 @@ def train(
         if show_train_summary:
             print(self._describe_round(self._round, self._summary))
 
+        # Avoid keeping the gradients in the resulting network, which can
+        # cause memory leakage when benchmarking.
+        self._neural_net.zero_grad()
```

Comment:
It's unclear to me what this line does. The gradients should be freed after the updates anyways. The following prints only `None` for me:
```python
import torch
from sbi import utils as utils
from sbi.inference import SNPE

num_dim = 3
prior = utils.BoxUniform(low=-2torch.ones(num_dim), high=2torch.ones(num_dim))

def simulator(parameter_set):
    result = 1.0 + parameter_set + torch.randn(parameter_set.shape) * 0.1
    return result

inference = SNPE(prior=prior)
theta = prior.sample((500,))
x = simulator(theta)
net = inference.append_simulations(theta, x).train(max_num_epochs=40)

for p in net.parameters():
    print(p.grad)
```

Reply:
Thanks for the reproducer. Indeed, I am also obtaining only `None`. I should have checked this.. :sweat_smile: 

So what is happening is that the `deepcopy` actually does not copy the gradient buffers. But putting `print(list(self._neural_net.parameters())[0].grad)` in line 354 shows that the gradient are not zero in `self._neural_net`. (`optimizer.step` does not set grad buffer to 0, else we would not need the `optimizer.zero_grad`).

I would say it is not critical but as the network is kept as an attribute of the `SNPE` instance, freeing the memory is still a good practice. If you think it is worth it, I will update the comment, else, I will remove. Let me know what you prefer.

Reply:
Ok, yes, I agree. For consistency, could you add the same line to `snle_base.py` and to `snre_base.py`?

Reply:
Indeed, I did this in the last commit.